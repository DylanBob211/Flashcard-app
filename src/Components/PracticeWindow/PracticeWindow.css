.practiceWindow_container {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  z-index: 2;
  border: solid #000 4px;
  border-radius: 20px;
  color: black;
  background: #ddf;
  min-width: 75%;
  min-height: 75%;
}

.practiceWindow_overlay {
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: rgba(0,0,0,.25);
  z-index: 2;
}