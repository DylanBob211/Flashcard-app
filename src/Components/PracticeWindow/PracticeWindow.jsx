import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import useOutsideClick from '../../Hooks/useOutsideHandler';
import './PracticeWindow.css';
import Flashcard from './Flashcard/Flashcard';
import ExerciseSettingsPanel from './ExerciseSettingsPanel/ExerciseSettingsPanel';

const PracticeWindow = ({ windowState, closeExerciseWindow }) => {
  const refWrapper = useRef(null);
  useOutsideClick(refWrapper, closeExerciseWindow);
  const practiceWindowWrapper = (Component, data) => <div ref={refWrapper} className="practiceWindow_container"><Component data={data} /></div>;

  switch (windowState.case) {
    case 'flashcard': return practiceWindowWrapper(Flashcard, windowState.data);
    case 'practise': return practiceWindowWrapper(ExerciseSettingsPanel, windowState.data);
    case '': return null;
    default: return null;
  }
};

PracticeWindow.propTypes = {
  windowState: PropTypes.shape({
    case: PropTypes.string,
    data: PropTypes.array || PropTypes.object, // TODO: non funziona, ma per capirci che qualcosa va fatto
  }).isRequired,
  closeExerciseWindow: PropTypes.func.isRequired,
};

export default PracticeWindow;
