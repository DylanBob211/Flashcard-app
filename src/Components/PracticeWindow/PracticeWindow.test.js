import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import testWrapper from '../../Helpers/testComponentMountWrapper';
import PracticeWindow from './PracticeWindow';

describe('PracticeWindow', () => {
  let mountedPracticeWindow;

  beforeEach(() => {
    mountedPracticeWindow = undefined;
    jest.clearAllMocks();
  });
  const props = {
    closeExerciseWindow: jest.fn(),
  };

  it('matches to snapshot', () => {
    expect(testWrapper(PracticeWindow, props, true)).toMatchSnapshot();
  });

  describe('when `windowState.case` is empty', () => {
    beforeEach(() => {
      mountedPracticeWindow = testWrapper(PracticeWindow, props);
    });

    it('does not render anything', () => {
      const divContainer = mountedPracticeWindow.find('.practiceWindow_container');
      expect(divContainer.exists()).toEqual(false);
    });

    describe('when something is rendered', () => {
      beforeEach(() => {
        props.windowState = {
          case: 'sth',
          data: [],
        };
        mountedPracticeWindow = testWrapper(PracticeWindow, props);
      });

      it.skip('calls the closeExerciseWindow prop when clicking outside', () => {
        const map = {};

        document.addEventListener = jest.fn((event, cb) => {
          map[event] = cb;
        });

        expect(props.closeExerciseWindow).toHaveBeenCalled();
      });

      it.skip('does not call the closeExerciseWindow prop when clicking inside', () => {
        const map = {};

        document.addEventListener = jest.fn((event, cb) => {
          map[event] = cb;
        });
        const props = {
          windowState: {
            case: 'sth',
            data: [],
          },
          closeExerciseWindow: jest.fn(),
        };
        const wrapper = mount(<PracticeWindow {...props} />);
        map.mousedown({
          // eslint-disable-next-line react/no-find-dom-node
          target: ReactDOM.findDOMNode(wrapper.instance()),
        });

        expect(props.closeExerciseWindow).not.toHaveBeenCalled();
      });

      describe('when a windowState object is passed with key `case` = `flashcard`', () => {
        let flashcardComponent;
        beforeEach(() => {
          props.windowState = {
            case: 'flashcard',
            data: [{ word: 'someData', url: [] }],
          };
          mountedPracticeWindow = testWrapper(PracticeWindow, props);
          flashcardComponent = mountedPracticeWindow
            .find('Flashcard');
        });

        it('renders a Flashcard component', () => {
          expect(flashcardComponent.exists()).toEqual(true);
        });

        it('passes `data` key reference to the Flashcard component', () => {
          expect(flashcardComponent.props().data).toEqual([{ word: 'someData', url: [] }]);
        });
      });

      describe('when a windowState object is passed with key `case` = `list', () => {
        beforeEach(() => {
          props.windowState = {
            case: 'list',
            data: ['someData'],
          };
          mountedPracticeWindow = testWrapper(PracticeWindow, props);
        });

        it('renders a composed ExerciseSettingsPanel component', () => {
          const panelComponent = mountedPracticeWindow
            .find('ExerciseSettingsPanel');
          expect(panelComponent.exists()).toEqual(true);
        });

        it('passes `data` key reference to the ExerciseSettingsPanel component', () => {
          const panelComponent = mountedPracticeWindow
            .find('ExerciseSettingsPanel');
          expect(panelComponent.props().data).toEqual(['someData']);
        });
      });
    });
  });
});
