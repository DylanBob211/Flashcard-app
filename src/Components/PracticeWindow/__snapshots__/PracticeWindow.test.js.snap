// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PracticeWindow matches to snapshot 1`] = `""`;
